server/api/roster/roster.controller.ts

```diff
@@ -175,71 +287,165 @@ class RosterController {
+      org: org.id,
+    })
+    .andWhere('unit like :name', { name: role.indexPrefix.replace('*', '%') });
+}
+
+async function getAllowedRosterColumns(org: Org, role: Role) {
+  const allColumns = await getRosterColumns(org.id);
+  const fineGrained = !(role.allowedRosterColumns.length === 1 && role.allowedRosterColumns[0] === '*');
+  return allColumns.filter(column => {
+    if (fineGrained && role.allowedRosterColumns.indexOf(column.name) < 0) {
+      return false;
+    }
+    if (!role.canViewPII && column.pii) {
+      return false;
+    }
+    return role.canViewPHI || !column.phi;
+  });
+}
+
+export async function getRosterColumns(orgId: number) {
+  const customColumns = (await CustomRosterColumn.find({
+    where: {
+      org: orgId,
+    },
+  })).map(customColumn => {
+    const columnInfo: RosterColumnInfo = {
+      ...customColumn,
+      displayName: customColumn.display,
+      custom: true,
+      updatable: true,
     };
+    return columnInfo;
+  });
+  return [...BaseRosterColumns, ...customColumns];
+}
 
-    return responseColumnInfo;
+function setCustomColumnFromBody(column: CustomRosterColumn, body: CustomColumnData) {
+  if (body.name) {
+    column.name = body.name;
+  }
+  if (body.displayName) {
+    column.display = body.displayName;
+  }
+  if (body.type) {
+    column.type = body.type;
+  }
+  if (body.pii != null) {
+    column.pii = body.pii;
+  }
+  if (body.phi != null) {
+    column.phi = body.phi;
+  }
+  if (body.required != null) {
+    column.required = body.required;
+  }
+}
+
+function setRosterParamsFromBody(entry: Roster, body: RosterEntryData, columns: RosterColumnInfo[], newEntry: boolean = false) {
+  columns.forEach(column => {
+    if (newEntry || column.updatable) {
+      getColumnFromBody(entry, body, column, newEntry);
+    }
   });
 }
 
-function setRosterParamsFromBody(entry: Roster, body: RosterEntryData) {
-  entry.firstName = getRequiredParam('firstName', body);
-  entry.lastName = getRequiredParam('lastName', body);
-  entry.unit = getRequiredParam('unit', body);
-  entry.billetWorkcenter = getRequiredParam('billetWorkcenter', body);
-  entry.contractNumber = getRequiredParam('contractNumber', body);
-  entry.rateRank = getOptionalParam('rateRank', body);
-  entry.pilot = getOptionalParam('pilot', body, 'boolean');
-  entry.aircrew = getOptionalParam('aircrew', body, 'boolean');
-  entry.cdi = getOptionalParam('cdi', body, 'boolean');
-  entry.cdqar = getOptionalParam('cdqar', body, 'boolean');
-  entry.dscacrew = getOptionalParam('dscacrew', body, 'boolean');
-  entry.advancedParty = getOptionalParam('advancedParty', body, 'boolean');
-  entry.pui = getOptionalParam('pui', body, 'boolean');
-  const covid19TestReturnDate = getOptionalParam('covid19TestReturnDate', body);
-  if (covid19TestReturnDate) {
-    entry.covid19TestReturnDate = new Date(covid19TestReturnDate);
+function getColumnFromBody(roster: Roster, row: RosterEntryData, column: RosterColumnInfo, newEntry: boolean) {
```

Comment:
There is a bit of a difference about how we set the value depending on if it's a custom column or not.  I could have it return the value, but then there will be a lot of duplication for 
```
    if (column.custom) {
      roster.customColumns[column.name] = objectValue;
    } else {
      Reflect.set(roster, column.name, objectValue);
    }
```

Reply:
I wonder if we could wrap that into a Roster function like `roster.setColumnValue(column, value)`. That way at least the setting would happen on the roster object itself, so that it's not being modified as an argument.

Reply:
We can come back to this though and potentially revisit all of those sorts of functions at once.